server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java

```diff
@@ -80,44 +100,91 @@ protected void doExecute(Task task, FieldCapabilitiesRequest request, final Acti
             return;
         }
 
-        final List<FieldCapabilitiesIndexResponse> indexResponses = Collections.synchronizedList(new ArrayList<>());
+        checkIndexBlocks(clusterState, concreteIndices);
+
+        final Map<String, FieldCapabilitiesIndexResponse> indexResponses = Collections.synchronizedMap(new HashMap<>());
         final FailureCollector indexFailures = new FailureCollector();
-        final Runnable countDown = createResponseMerger(request, totalNumRequest, indexResponses, indexFailures, listener);
 
+        // If all nodes are on version 7.16 or higher, then we group the shard requests and send a single request per node.
+        // Otherwise, for backwards compatibility we follow the old strategy of sending a separate request per shard.
+        final Map<String, List<ShardId>> shardsByNode;
+        final CountDown completionCounter;
+        if (clusterState.getNodes().getMinNodeVersion().onOrAfter(Version.V_7_16_0)) {
+            shardsByNode = groupShardsByNode(request, clusterState, concreteIndices, indexFailures);
+            completionCounter = new CountDown(shardsByNode.size() + remoteClusterIndices.size());
+        } else {
+            shardsByNode = null;
+            completionCounter = new CountDown(concreteIndices.length + remoteClusterIndices.size());
+        }
+
+        final Runnable countDown = createResponseMerger(request, completionCounter, indexResponses, indexFailures, listener);
         if (concreteIndices.length > 0) {
             // fork this action to the management pool as it can fan out to a large number of child requests that get handled on SAME and
             // thus would all run on the current transport thread and block it for an unacceptable amount of time
             // (particularly with security enabled)
             threadPool.executor(ThreadPool.Names.MANAGEMENT).execute(ActionRunnable.wrap(listener, l -> {
-                for (String index : concreteIndices) {
-                    new TransportFieldCapabilitiesIndexAction.AsyncShardsAction(
-                        transportService,
-                        clusterService,
-                        prepareLocalIndexRequest(request, index, localIndices, nowInMillis),
-                        new ActionListener<FieldCapabilitiesIndexResponse>() {
-                            @Override
-                            public void onResponse(FieldCapabilitiesIndexResponse result) {
-                                if (result.canMatch()) {
-                                    indexResponses.add(result);
+                if (shardsByNode != null) {
+                    for (Map.Entry<String, List<ShardId>> entry : shardsByNode.entrySet()) {
+                        String nodeId = entry.getKey();
+                        List<ShardId> shardIds = entry.getValue();
+
+                        DiscoveryNode node = clusterState.getNodes().get(nodeId);
+                        assert node != null;
```

Comment:
Can `node` be `null`? I didn't think so, but `TransportFieldCapabilitiesIndexAction` has some logic to handle this case: https://github.com/elastic/elasticsearch/blob/6a88d848315a4d03b92412321ee1ca578c006afb/server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesIndexAction.java#L244

Reply:
I don't think it's a valid assertion. If the node is null, we should fail the request the same way since the node may vanished at any time. In the current model, that's ok because we'd try another replica but the new model requires to try these shards on potentially more than node. So unless we consider that we don't want to retry on replica, we'll need to handle failures differently.

Reply:
We had a conversation offline where we decided the current strategy (where we don't retry failures) was not acceptable. For example if a shard is being relocated, the coordinator could have an out-of-date view of where it is, and the field caps request would fail. It's challenging for clients to handle these partial failures.

Maybe in an immediate follow-up, we can add support for a retry step, where we do another round of node requests to retry on other shards/ replicas. Like the other one around reducing responses, this follow-up would need to ship in the same release. For now in this PR I'd just replace the incorrect assertion with an error.